Fixes service controller update race condition

jhorwit2 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

Fixes service controller update race condition that can happen with the node sync loop and the worker(s). This PR allows the node sync loop to utilize the same work queue as service updates so that the queue can ensure the service is being acted upon by only one goroutine. 

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #53462

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
NONE
```

/cc @wlan0 @luxas @prydie @andrewsykim 

/sig cluster-lifecycle
/area cloudprovider

jhorwit2 committed:
Fixes service controller update race condition

jhorwit2 committed:
generated files

jhorwit2 commented:
This is WIP because I still need to run more tests in an actual cluster. Functionally this is ready for review though.

wlan0 commented on pkg/controller/service/service_controller.go:
```diff
@@ -435,6 +446,8 @@ func (s *serviceCache) delete(serviceName string) {
 }
 
+// needsUpdate checks to see if there were any changes between the old and new service that would require a load balancer update.
+// This method does not and should not check if the hosts have changed.
```
:+1:

wlan0 commented:
Thanks for this! This LGTM

thockin commented:
Assign back to me once it is LGTM'ed and ready for approval.  I give @wlan0 the review on this one.

jhorwit2 commented:
@wlan0 I ran some tests on an existing cluster and the results were good. I removed the WIP label. 


First test was cordoning a node. As the logs indicate UpdateLoadBalancer was called instead of EnsureLoadBalancer.  
```
I1110 21:16:05.957196       7 service_controller.go:666] Detected change in list of current cluster nodes. New node set: [host1 host2 host3 host4]
I1110 21:16:05.957297       7 load_balancer.go:406] Attempting to update load balancer 'loadbalancer-service-1'
I1110 21:16:07.067208       7 load_balancer.go:286] Applying "update" action on backend set `TCP-443` for lb `loadbalancer-1-ocid`
I1110 21:16:32.780236       7 load_balancer.go:286] Applying "update" action on backend set `TCP-80` for lb `loadbalancer-1-ocid`
I1110 21:16:47.265296       7 load_balancer.go:217] Successfully ensured load balancer "loadbalancer-service-1"
I1110 21:16:47.265401       7 load_balancer.go:406] Attempting to update load balancer 'loadbalancer-service-2
I1110 21:16:47.265471       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-internal", UID:"171b9fe9-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"15751", FieldPath:""}): type: 'Normal' reason: 'UpdatedLoadBalancer' Updated load balancer with new hosts
I1110 21:16:48.458252       7 load_balancer.go:286] Applying "update" action on backend set `TCP-443` for lb `loadbalancer-2-ocid`
I1110 21:17:10.260149       7 load_balancer.go:286] Applying "update" action on backend set `TCP-80` for lb `loadbalancer-2-ocid`
I1110 21:17:21.425777       7 load_balancer.go:217] Successfully ensured load balancer "k8s-us-phx-a-1748042b-bb30-11e7-bf4c-0000170092e3"
I1110 21:17:21.425913       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-public", UID:"1748042b-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"16082", FieldPath:""}): type: 'Normal' reason: 'UpdatedLoadBalancer' Updated load balancer with new hosts
```

The next test I ran was I uncordoned the same node. Once I saw the update start happening I updated the service. As expected, they were processed in order based on the workqueue. EnsureLoadBalancer was called for the service update and UpdateLoadBalancer for updating the two load balancers for the node change.  

```
I1110 21:19:25.957522       7 service_controller.go:666] Detected change in list of current cluster nodes. New node set: [host1 k8swa-cec30922.ad2.prd.us-phx.odc.im host2 host3 host4]
I1110 21:19:25.957635       7 load_balancer.go:406] Attempting to update load balancer 'loadbalancer-service-1'
I1110 21:19:27.038952       7 load_balancer.go:286] Applying "update" action on backend set `TCP-443` for lb `loadbalancer-1-ocid`
I1110 21:19:43.324369       7 load_balancer.go:286] Applying "update" action on backend set `TCP-80` for lb `loadbalancer-1-ocid`
I1110 21:19:59.128539       7 load_balancer.go:217] Successfully ensured load balancer "loadbalancer-service-1"
I1110 21:19:59.128699       7 load_balancer.go:406] Attempting to update load balancer 'loadbalancer-service-2
I1110 21:19:59.128740       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-internal", UID:"171b9fe9-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"15751", FieldPath:""}): type: 'Normal' reason: 'UpdatedLoadBalancer' Updated load balancer with new hosts
I1110 21:20:00.264083       7 load_balancer.go:286] Applying "update" action on backend set `TCP-443` for lb `loadbalancer-2-ocid`
I1110 21:20:11.254631       7 load_balancer.go:286] Applying "update" action on backend set `TCP-80` for lb `loadbalancer-2-ocid`
I1110 21:20:20.577953       7 load_balancer.go:217] Successfully ensured load balancer "k8s-us-phx-a-1748042b-bb30-11e7-bf4c-0000170092e3"
I1110 21:20:20.578069       7 service_controller.go:309] Ensuring LB for service test/ingress-controller-internal
I1110 21:20:20.578137       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-public", UID:"1748042b-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"16082", FieldPath:""}): type: 'Normal' reason: 'UpdatedLoadBalancer' Updated load balancer with new hosts
I1110 21:20:20.578211       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-internal", UID:"171b9fe9-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"3874587", FieldPath:""}): type: 'Normal' reason: 'EnsuringLoadBalancer' Ensuring load balancer
I1110 21:20:21.700978       7 load_balancer.go:286] Applying "update" action on backend set `TCP-443` for lb `loadbalancer-1-ocid`
I1110 21:20:47.476569       7 load_balancer.go:352] Applying "delete" action on listener `TCP-80` for lb `loadbalancer-1-ocid`
I1110 21:21:06.944680       7 load_balancer.go:286] Applying "delete" action on backend set `TCP-80` for lb `loadbalancer-1-ocid`
I1110 21:21:22.450553       7 load_balancer.go:286] Applying "create" action on backend set `TCP-8080` for lb `loadbalancer-1-ocid`
I1110 21:21:47.989922       7 load_balancer.go:352] Applying "create" action on listener `TCP-8080` for lb `loadbalancer-1-ocid`
I1110 21:22:12.294323       7 load_balancer.go:217] Successfully ensured load balancer "loadbalancer-service-1"
I1110 21:22:12.294379       7 service_controller.go:334] Not persisting unchanged LoadBalancerStatus for service test/ingress-controller-internal to registry.
I1110 21:22:12.294470       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"test", Name:"ingress-controller-internal", UID:"171b9fe9-bb30-11e7-bf4c-0000170092e3", APIVersion:"v1", ResourceVersion:"3874587", FieldPath:""}): type: 'Normal' reason: 'EnsuredLoadBalancer' Ensured load balancer
```

liggitt commented on pkg/controller/service/service_controller.go:
```diff
@@ -241,6 +240,20 @@ func (s *ServiceController) processServiceUpdate(cachedService *cachedService, s
+		if !s.needsUpdate(cachedService.state, service) {
+			// The service does not require an update which means it was placed on the work queue
+			// by the node sync loop and indicates that the hosts need to be updated.
+			err := s.updateLoadBalancerHosts(service)
```
Inferring source from state seems a little odd… does doing host updates when needsUpdate returns false mean this controller will do work on every service even when it should be in steady state?

jhorwit2 replied:
Services are only added to the queue under three conditions:
1. The service requires an update. That's determined by `needsUpdate` which is checked in the informer OnUpdate.
2. The node sync loop determined that hosts change from the last sync.
3. An error occurred so retry.

The `UpdateLoadBalancer` method is supposed to be cheap for cloud providers. It was added so that cloud providers could have a method that handled _only_ updating load balancer hosts. 

This approach should not (and doesn't from what I tell in my tests) add any extra calls from before.

MrHohn replied:
From https://github.com/kubernetes/kubernetes/issues/56443. I think the logic here is too hacky and may not cover all corner cases. What we really want here is to distinguish "service update" and "nodeSync update". The condition `!s.needsUpdate(cachedService.state, service)` is too broad that it includes also the retry case of "service update" (given how we cache service).

Besides, putting "nodeSync update" into the same work queue as "service update" might introduce another problem that one update could override the other. Ref https://github.com/kubernetes/kubernetes/issues/52495#issuecomment-330670308, within the working queue we don't save duplicate key, if both "nodeSync update" and  "service update" come in before anyone leaves the queue, it will end up with only one update (depends on how we decide what update it is). It seems to me that the working queue mechanism also needs to be adjusted before we can put in "nodeSync update".

cc @bowei

jhorwit2 replied:
> Besides, putting "nodeSync update" into the same work queue as "service update" might introduce another problem that one update could override the other

One update won't overwrite the other because the sync checks if the service needs an update. Both cloud provider calls (`EnsureLoadBalancer` and `UpdateLoadBalancer`) will update the hosts, so if both happen it will go with `EnsureLoadBalancer` which is what we want.

jhorwit2 replied:
As I mentioned in the other PR https://github.com/kubernetes/kubernetes/pull/56448#discussion_r153508837, the finalizer support will clean this all up and I think we should revert this until the finalizer PR cleans up all the cache/delete logic.

MrHohn replied:
> Both cloud provider calls (EnsureLoadBalancer and UpdateLoadBalancer) will update the hosts, so if both happen it will go with EnsureLoadBalancer

I think this assumption is inaccurate, ref the LoadBalancer interface, it doesn't explicitly define `EnsureLoadBalancer()` should update the hosts:
https://github.com/kubernetes/kubernetes/blob/e5aec8645d740123dfc8f1400b66bca05673b833/pkg/cloudprovider/cloud.go#L100-L109

And in fact, `EnsureLoadBalancer()` in GCE cloudprovider doesn't update hosts. Hence for GCE this is the case where "service update" overrides "nodeSync update".

jhorwit2 replied:
> creates a new load balancer 'name', or updates the existing one

Should include the backends (nodes). Why would it partially update the load balancer?

jhorwit2 replied:
@MrHohn from what I see GCE does make sure that backends are up-to-date on EnsureLoadBalancer calls via [this method](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L408), which is called by [`ensureInternalLoadBalancer`](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_internal.go#L37) and `ensureExternalLoadBalancer` does as well via chekcing first [this method](https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L675).

MrHohn replied:
Thanks for gathering the links, I didn't look into the internal one before, seems like it does check for hosts update. Though ATM the external one doesn't check for hosts update.
https://github.com/kubernetes/kubernetes/blob/e57accbd03b1265340fdd266d0afe2b7a8e1c171/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L675-L676

jhorwit2 replied:
Ah, missed that comment. The external one is definitely more complex than the internal one 👼 

It sounds like we should definitely revert this then and we need to come to an agreement on what `EnsureLoadBalancer` and `UpdateLoadBalancer` should do for each cloud provider. It was my understanding that `EnsureLoadBalancer` should completely update the load balancer, which is how AWS, Azure, Oracle & DigitalOcean handle it (only ones i checked). cc @wlan0 @luxas 

I'll open a PR to revert this for 1.9 @MrHohn

liggitt commented on pkg/controller/service/service_controller_test.go:
```diff
@@ -233,30 +258,44 @@ func TestUpdateNodesInExternalLoadBalancer(t *testing.T) {
 		},
-		{
-			// One service has an external load balancer and one is nil: one call.
-			services: []*v1.Service{
```
Why was this case removed?

jhorwit2 replied:
It's not possible for a service to be nil when `updateLoadBalancerHosts` is called and if it was nil the controller would panic way before that method got called in the `syncService` flow. I didn't see it valuable to add a nil check here because then i should also add it to `createLoadBalancerIfNeeded` and other methods.

wlan0 commented:
Other than the concerns introduced by @liggitt, I'm also wondering if there will be a noticeable delay in updating hosts for LBs now, since we are adding it back to the queue, instead of processing it right away. 

Is that something to consider, or is it too insignificant?

jhorwit2 commented:
@wlan0 Prior to this PR if a call to `UpdateLoadBalancer` failed then the service would be added to the [`servicesToUpdate`](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/service/service_controller.go#L664) list on the controller object. It wasn't retried immediately and only retried the next sync loop which is every 100s if the nodes didn't changes so retries occur faster this way.

wlan0 commented:
I meant the normal case, not just the failure case. But your answer gave me the information.

>  It wasn't retried immediately and only retried the next sync loop which is every 100s so retries occur faster this way.

If that delay was acceptable, this will be fine.

jhorwit2 commented:
@liggitt PTAL

liggitt commented:
no other comments from me, will defer to @kubernetes/sig-scheduling-pr-reviews

dims commented:
@enisoc Can you please mark this for 1.9 as the issue it fixes (53462) is marked as such

dims commented:
@enisoc looks like we need a `status/approved-for-milestone` as well

enisoc commented:
@dims The owning SIG(s) should decide if they approve of this change going into the release.

jhorwit2 commented:
@luxas could you please add `status/approved-for-milestone`?

luxas commented:
ping @wlan0 for a re-review
/assign @thockin 
for approval

thockin commented:
I can't take this review right now.  Assign back to me once it is LGTM'ed and ready for approval.

jhorwit2 commented:
/status in-progress

jhorwit2 commented:
@wlan0 PTAL

wlan0 commented:
This LGTM. I don't see any changes since my last review.

jhorwit2 commented:
Thanks!

/assign @thockin 
For approval.

thockin commented:
@wlan0 you have to say `/lgtm` :)

/lgtm
/approve

jhorwit2 commented:
/test pull-kubernetes-unit

jberkus commented:
/priority critical-urgent

/remove-priority important-longterm

adjusting priorities for code freeze
